src/libcretonne/cfg.rs

```diff
@@ -0,0 +1,270 @@
+//! A control flow graph represented as mappings of extended basic blocks to their predecessors.
+//! Predecessors are denoted by tuples of EBB and branch/jump instructions. Each predecessor
+//! tuple corresponds to the end of a basic block.
+//!
+//!```c
+//!     Ebb0:
+//!         ...          ; beginning of basic block
+//!
+//!         ...
+//!
+//!         brz vx, Ebb1 ; end of basic block
+//!
+//!         ...          ; beginning of basic block
+//!
+//!         ...
+//!
+//!         jmp Ebb2     ; end of basic block
+//!```
+//!
+//! Here Ebb1 and Ebb2 would each have a single predecessor denoted as (Ebb0, `brz vx, Ebb1`)
+//! and (Ebb0, `jmp Ebb2`) respectively.
+
+use repr::Function;
+use entities::{Inst, Ebb};
+use instructions::InstructionData;
+use std::collections::{BTreeSet, BTreeMap, btree_map};
+
+/// A basic block denoted by its enclosing Ebb and last instruction.
+pub type Predecessor = (Ebb, Inst);
+
+/// Storing predecessors in a BTreeSet ensures that their ordering is
+/// stable with no duplicates.
+pub type PredecessorSet = BTreeSet<Predecessor>;
+
+/// The Control Flow Graph maintains a mapping of ebbs to their predecessors
+/// where predecessors are basic blocks.
+#[derive(Debug)]
+pub struct ControlFlowGraph {
+    data: BTreeMap<Ebb, PredecessorSet>,
+}
+
+impl ControlFlowGraph {
+
+    /// During initialization mappings will be generated for any existing
+    /// blocks within the CFG's associated function. Basic sanity checks will
+    /// also be performed to ensure that the blocks are well formed.
+    pub fn new(func: &Function) -> Result<ControlFlowGraph, String> {
+        let mut cfg = ControlFlowGraph{data: BTreeMap::new()};
+
+        // Even ebbs without predecessors should show up in the CFG, albeit
+        // with no entires.
+        for ebb in func.ebbs_numerically() {
+            try!(cfg.init_ebb(ebb));
+        }
+
+        for ebb in func.ebbs_numerically() {
+            // Flips to true when a terminating instruction is seen. So that if additional
+            // instructions occur an error may be returned.
+            let mut terminated = false;
+            for inst in func.ebb_insts(ebb) {
+                if terminated {
+                    return Err(format!("{} contains unreachable instructions.", ebb));
```

Comment:
This is correct, but I think we should leave this check to the verifier. It doesn't actually affect whether we can compute a control flow graph or not.